Code/src/main/java/nl/utwente/group10/haskell/expr/Function.java

```diff
@@ -0,0 +1,167 @@
+package nl.utwente.group10.haskell.expr;
+
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellException;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
+import nl.utwente.group10.haskell.type.FuncT;
+import nl.utwente.group10.haskell.type.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Represents a user-defined Haskell function. Holds arguments which can be used for building the function.
+ */
+public class Function extends Expr {
+
+    /**
+     * An argument for a function. Can be used as a variable in a function definition.
+     */
+    public class FunctionArgument extends Expr {
+
+        /** The function this argument is for. */
+        private Function function;
+
+        /**
+         * @param function The function this argument is for.
+         */
+        protected FunctionArgument(Function function) {
+            this.function = function;
+        }
+
+        /**
+         * @return The 0-indexed position of this argument in the function definition.
+         */
+        public int getPosition() {
+            return this.function.arguments.indexOf(this);
+        }
+
+        @Override
+        public Type analyze(Env env, GenSet genSet) throws HaskellException {
+            return HindleyMilner.makeVariable(); // TODO Make sure that this is correctly unified and/or make this editable.
+        }
+
+        @Override
+        public String toHaskell() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
```

Comment:
This will likely not be unique when lambda's get nested somehow. I don't see how they would, but I imagine they could, and that could lead to some human confusion:

    (\arg0 -> map (\arg0 -> arg0) arg0) 

If you drop the function's hashCode in there somewhere, you'll have a way of differentiating all those arg0's.